services/incentivisation/service.go

```diff
@@ -0,0 +1,518 @@
+	JoiningSession uint32
+	ActiveSession  uint32
+	Active         bool
+}
+
+func formatEnodeURL(publicKey string, ip string, port uint16) string {
+	return fmt.Sprintf("enode://%s:%s:%d", publicKey, ip, port)
+}
+
+func (n *Enode) toEnodeURL() string {
+	return formatEnodeURL(n.PublicKeyString(), n.IP.String(), n.Port)
+}
+
+func (n *Enode) PublicKeyString() string {
+	return hex.EncodeToString(n.PublicKey)
+}
+
+type Whisper interface {
+	Post(ctx context.Context, req whisper.NewMessage) (hexutil.Bytes, error)
+	NewMessageFilter(req whisper.Criteria) (string, error)
+	AddPrivateKey(ctx context.Context, privateKey hexutil.Bytes) (string, error)
+	DeleteKeyPair(ctx context.Context, key string) (bool, error)
+	GenerateSymKeyFromPassword(ctx context.Context, passwd string) (string, error)
+	GetFilterMessages(id string) ([]*whisper.Message, error)
+}
+
+type ServiceConfig struct {
+	RPCEndpoint     string
+	ContractAddress string
+	IP              string
+	Port            uint16
+}
+
+type Service struct {
+	w               Whisper
+	whisperKeyID    string
+	whisperSymKeyID string
+	whisperFilterID string
+	nodes           map[string]*Enode
+	ticker          *time.Ticker
+	quit            chan struct{}
+	config          *ServiceConfig
+	contract        Contract
+	privateKey      *ecdsa.PrivateKey
+	log             log.Logger
+	// The first round we will not be voting, as we might have incomplete data
+	initialSession uint64
+	// The current session
+	currentSession uint64
+	whisperPings   map[string][]uint32
+}
+
+// New returns a new Service
+func New(prv *ecdsa.PrivateKey, w Whisper, config *ServiceConfig, contract Contract) *Service {
+	logger := log.New("package", "status-go/incentivisation/service")
+	return &Service{
+		w:            w,
+		config:       config,
+		privateKey:   prv,
+		log:          logger,
+		contract:     contract,
+		nodes:        make(map[string]*Enode),
+		whisperPings: make(map[string][]uint32),
+	}
+}
+
+// Protocols returns a new protocols list. In this case, there are none.
+func (s *Service) Protocols() []p2p.Protocol {
+	return []p2p.Protocol{}
+}
+
+// APIs returns a list of new APIs.
+func (s *Service) APIs() []rpc.API {
+	s.log.Info("APIS CALLED")
+	apis := []rpc.API{
+		{
+			Namespace: "incentivisation",
+			Version:   "1.0",
+			Service:   NewAPI(s),
+			Public:    true,
+		},
+	}
+	return apis
+}
+
+// checkRegistered checks that a node is registered with the contract
+func (s *Service) checkRegistered() error {
+	registered, err := s.registered()
+	if err != nil {
+		s.log.Error("error querying contract", "registered", err)
+		return err
+	}
+
+	if registered {
+		s.log.Info("Already registered")
+		return nil
+	}
+	s.log.Info("not registered")
+	_, err = s.register()
+	if err != nil {
+		s.log.Error("error querying contract", "registered", err)
+		return err
+	}
+	return nil
+}
+
+// newSession checks if we are in a new session and updates the session if so
+func (s *Service) newSession() (bool, error) {
```

Comment:
maybe `ensureSession` or `getSession` would better describe what the method does?